File: src/order/Turn.java
package order;


import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class Turn implements Serializable {
    private int workDay;
    private int dayTime;
    private List<Student> students;
    private List<Student> fixedStudents;
    static int maxStudentCount = 0;


    Turn(int workDay, int dayTime) {
        this.workDay = workDay;
        this.dayTime = dayTime;
        students = new ArrayList<>();
        fixedStudents = new ArrayList<>();
    }


    /**
     * 获取当前班次所有学生的空闲班次的总和
     *
     * @return 学生们所剩空闲时间总和
     */
    int getAllStudentCanCount() {
        int n = 0;
        for (Student s : students) {
            n += s.getFreeTimeCount();
        }
        return n;
    }


    void addStudengt(Student student) {
        students.add(student);
        student.addSuitableTurn(this);
    }

    public void reset() {
        maxStudentCount = 0;
        for (int i = 0, j = students.size(); i < j; i++) {
            students.remove(0);
        }
        for (int i = 0, j = fixedStudents.size(); i < j; i++) {
            fixedStudents.remove(0);
        }
    }

    /**
     * 把班次和学生进行互相绑定
     *
     * @param student 需要绑定的学生
     */
    void addFixedStudengt(Student student) {

        if (!isFixedExists(student)) {
            student.fixed(this);
            fixedStudents.add(student);
        }
        if (student.getFixedCount() > 2) {
            System.out.println();
        }
    }

    private boolean isFixedExists(Student student) {
        boolean n = false;
        for (Student s : fixedStudents) {
            if (student == s) {
                n = true;
            }
        }
        return n;
    }

    int getStudentNumber() {
        return students.size();
    }

    List<Student> getStudents() {
        return students;
    }

    public int getDayTime() {
        return dayTime;
    }

    public int getWorkDay() {
        return workDay;
    }



    int getFixedStudentCount() {
        return fixedStudents.size();
    }

    public List<Student> getFixedStudents() {
        return fixedStudents;
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder
                .append("第")
                .append(dayTimeChange(this))
                .append("节课：");
        return stringBuilder.toString();
    }


    private String dayTimeChange(Turn turn) {
        switch (turn.getDayTime()) {
            case 1:
                return "一二";
            case 2:
                return "三四";
            case 3:
                return "五六";
            case 4:
                return "七八";
        }
        return "发生错误";
    }
}

File: src/order/Turn.kt
package order

import java.io.Serializable

/**
 * @author Jovines
 * @create 2020-04-13 3:23 下午
 * 描述:
 * @param column 该轮次在多少列，从1开始
 * @param row 该轮次在多少行，从1开始
 */
class Turn(val row: Int, val column: Int) : Serializable {

    val people = ArrayList<Item>()
    val fixedPeople = ArrayList<Item>()

    /**
     * 获取当前班次所有Item的空闲班次的总和
     *
     * @return Item们所剩空闲时间总和
     */
    val allItemCanCount: Int
        get() {
            var n = 0
            for (s in people) {
                n += s.freeTimeCount
            }
            return n
        }

    fun addStudent(item: Item) {
        people.add(item)
        item.addSuitableTurn(this)
    }


    fun reset() {
        maxItemCount = 0
        people.clear()
        fixedPeople.clear()
    }

    /**
     * 把班次和Item进行互相绑定
     *
     * @param item 需要绑定的Item
     */
    fun addFixedStudent(item: Item) {
        if (!isFixedExists(item)) {
            item.fixed(this)
            fixedPeople.add(item)
        }
        if (item.fixedCount > 2) {
            println()
        }
    }

    private fun isFixedExists(item: Item): Boolean {
        var n = false
        for (s in fixedPeople) {
            if (item == s) {
                n = true
            }
        }
        return n
    }

    val itemNumber: Int
        get() = people.size


    val fixedItemCount: Int
        get() = fixedPeople.size


    override fun toString(): String {
        val stringBuilder = StringBuilder()
        stringBuilder.append("${row}行${column}列：")
        for (s in fixedPeople) {
            stringBuilder.append(s.name + " ")
        }
        return stringBuilder.toString()
    }

    companion object {
        var maxItemCount = 0
    }
}